### PR TITLE
Fine-grained error types - capi edition

### PR DESCRIPTION
**Pull request by robertbastian**

### CODE REVIEW

File: ffi/capi/src/errors.rs
Comment (Manishearth) [inline ffi/capi/src/errors.rs:59]: question: do we have plans for this, or is this the permanent name?

If these errors are permanently ZSTs I think we should just return `()`, yes?

Comment (Manishearth) [reply ffi/capi/src/errors.rs:59]: Okay

**Review by Manishearth — DISMISSED** [review]
r+ with the understanding that this makes Diplomat ZSTs a 2.0 blocker

**Review by Manishearth — APPROVED** [review]

**Comment by robertbastian** [comment]
@sffc you previously voiced an opinion on this, do you want to review?

**Review by sffc — COMMENTED** [review]
Seems about right. `TodoZst` seems weird (why not just make it an enum) but I defer to @Manishearth on that and a more detailed review

File: ffi/capi/src/errors.rs
Comment (robertbastian) [reply ffi/capi/src/errors.rs:61]: https://github.com/rust-diplomat/diplomat/pull/494 adds support, using it is blocked on JS2

**Pull request merged**
